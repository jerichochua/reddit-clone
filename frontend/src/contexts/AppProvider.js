import React, { createContext, useContext, useReducer } from 'react';
import jwt_decode from 'jwt-decode';

const AppContext = createContext();
AppContext.displayName = 'AppContext';

const token = localStorage.getItem('token');
const userId = token ? jwt_decode(token).id : null;
const username = token ? jwt_decode(token).username : null;

const initialState = {
  userId,
  username,
  token,
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_USER':
      return {
        ...state,
        userId: action.payload.userId,
        username: action.payload.username,
      };
    case 'SET_TOKEN':
      return { ...state, token: action.payload };
    case 'LOGOUT':
      return { ...state, userId: null, username: null, token: null };
    default:
      return state;
  }
};

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <AppContext.Provider value={{ state, dispatch }}>
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  return useContext(AppContext);
};
