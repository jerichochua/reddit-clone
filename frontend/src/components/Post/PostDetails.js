import React from 'react';
import getRelativeTime from '../../util/getRelativeTime';
import PostButton from './PostButton/PostButton';
import { useAppContext } from '../../contexts/AppProvider';
import Votes from '../Vote/Votes';

const PostDetails = (props) => {
  const { state } = useAppContext();

  const isAuthor = state.username === props.post.author;

  return (
    <div className='post-details-container'>
      <Votes score={props.post.score} />
      <div className='post-content-container'>
        <div className='post-title'>
          <span>{props.post.title}</span>
        </div>
        <div className='post-content'>
          {props.post.post_type === 'text' && <p>{props.post.content}</p>}
          {props.post.post_type === 'link' && (
            <a href={props.post.post_url} target='_blank' rel='noreferrer'>
              {props.post.post_url}
            </a>
          )}
        </div>
        <div className='post-details'>
          <p>
            {props.post.comments} comment
            {parseInt(props.post.comments) === 1 ? '' : 's'} by{' '}
            {props.post.author} {getRelativeTime(props.post.created_at)}
          </p>
          {isAuthor && (
            <PostButton label='Delete' type='danger' onClick={props.onDelete} />
          )}
        </div>
      </div>
    </div>
  );
};

export default PostDetails;
