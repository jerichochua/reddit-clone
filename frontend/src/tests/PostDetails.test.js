import renderer from 'react-test-renderer';
import PostDetails from '../components/Post/PostDetails';
import { AppProvider } from '../contexts/AppProvider';

const post = {
  author: 'test',
  title: 'test',
  content: 'test',
  score: 1,
  comments: 1,
  created_at: '2020-01-01T00:00:00.000Z',
};

it('renders correctly', () => {
  const tree = renderer.create(
    <AppProvider>
      <PostDetails post={post} />
    </AppProvider>
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it('renders correctly with delete button', () => {
  const tree = renderer.create(
    <AppProvider value={{ username: 'test' }}>
      <PostDetails post={post} />
    </AppProvider>
  );
  expect(tree.toJSON()).toMatchSnapshot();
  expect(tree.root.findByProps({ label: 'Delete' })).toBeTruthy();
});

it('calls onDelete when delete button is clicked', () => {
  const onDelete = jest.fn();
  const tree = renderer.create(
    <AppProvider value={{ username: 'test' }}>
      <PostDetails post={post} onDelete={onDelete} />
    </AppProvider>
  );
  tree.root.findByProps({ label: 'Delete' }).props.onClick();
  expect(onDelete).toHaveBeenCalled();
});
