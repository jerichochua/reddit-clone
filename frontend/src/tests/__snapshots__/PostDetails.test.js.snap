// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="post-details-container"
>
  <div
    className="votes-container"
  >
    <button
      className="upvote-btn"
      onClick={[Function]}
    />
    <span>
      1
    </span>
    <button
      className="downvote-btn"
      onClick={[Function]}
    />
  </div>
  <div
    className="post-content-container"
  >
    <div
      className="post-title"
    >
      <span>
        test
      </span>
    </div>
    <div
      className="post-content"
    >
      <p>
        test
      </p>
    </div>
    <div
      className="post-details"
    >
      <p>
        1
         comment
         by
         
        test
         
        3 years ago
      </p>
    </div>
  </div>
</div>
`;

exports[`renders correctly with delete button 1`] = `
<div
  className="post-details-container"
>
  <div
    className="votes-container"
  >
    <button
      className="upvote-btn"
      onClick={[Function]}
    />
    <span>
      1
    </span>
    <button
      className="downvote-btn"
      onClick={[Function]}
    />
  </div>
  <div
    className="post-content-container"
  >
    <div
      className="post-title"
    >
      <span>
        test
      </span>
    </div>
    <div
      className="post-content"
    >
      <p>
        test
      </p>
    </div>
    <div
      className="post-details"
    >
      <p>
        1
         comment
         by
         
        test
         
        3 years ago
      </p>
      <button
        className="post-button danger"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
